### DISCUSSION TITLE
Error: Multiple 'Z' axis dims were found in this dataset. Please drop the unused dimension(s) before performing grid operations.

### DISCUSSION THREAD

**Discussion by mzelinka** [xCDAT/xcdat#608] [2. Questions]
### Question criteria

- [X] I added a descriptive title here.
- [X] I searched the [xCDAT GitHub Discussions](https://github.com/xCDAT/xcdat/discussions) to find a similar question and didn't find it.
- [X] I searched the [xCDAT documentation](https://xcdat.readthedocs.io/en/stable/) and/or [Xarray documentation](https://docs.xarray.dev/en/stable/).

### Describe your question

I am reading in a dataset and trying to regrid it.  I am getting the error noted in the title, despite the fact that I am explicitly setting only one coordinate to be the 'Z' axis.

### Are there are any possible answers you came across?

No.

### Minimal Complete Verifiable Example (MVCE)

```python
import xcdat as xc
import numpy as np

# Create a grid
LAT = np.arange(-89,91,2.0)
LON = np.arange(1.25,360,2.5)
output_grid = xc.regridder.grid.create_grid(lat=LAT, lon=LON)

# Read in data
path='/p/css03/esgf_publish/CMIP6/CFMIP/CNRM-CERFACS/CNRM-CM6-1/amip-p4K/r1i1p1f2/CFmon/clisccp/gr/v20190820/'
fn = path+'clisccp_CFmon_CNRM-CM6-1_amip-p4K_r1i1p1f2_gr_197901-201412.nc'
ds=xc.open_mfdataset(fn)

# Explicitly set plev7,tau coords to be 'Z' and 'tau'
ds.plev7.attrs['axis'] = 'Z'
ds.tau.attrs['axis'] = 'tau' # could use anything here, the error below still occurs

# Regrid
output = ds.regridder.horizontal('clisccp', output_grid, tool="xesmf", method="bilinear", extrap_method="inverse_dist")
```

### Relevant log output

```python
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In [34], line 13
     11 ds.tau.attrs['axis'] = None
     12 # Regrid:
---> 13 output = ds.regridder.horizontal('clisccp', output_grid, tool="xesmf", method="bilinear", extrap_method="inverse_dist")

File ~/miniconda3/envs/xclim2/lib/python3.10/site-packages/xcdat/regridder/accessor.py:322, in RegridderAccessor.horizontal(self, data_var, output_grid, tool, **options)
    317 except KeyError as e:
    318     raise ValueError(
    319         f"Tool {e!s} does not exist, valid choices {list(HORIZONTAL_REGRID_TOOLS)}"
    320     )
--> 322 input_grid = _get_input_grid(self._ds, data_var, ["X", "Y"])
    323 regridder = regrid_tool(input_grid, output_grid, **options)
    324 output_ds = regridder.horizontal(data_var, self._ds)

File ~/miniconda3/envs/xclim2/lib/python3.10/site-packages/xcdat/regridder/accessor.py:445, in _get_input_grid(ds, data_var, dup_check_dims)
    442 input_grid = ds.drop_dims(to_drop)
    444 # drops extra dimensions on input grid
--> 445 grid = input_grid.regridder.grid
    447 # preserve mask on grid
    448 if "mask" in ds:

File ~/miniconda3/envs/xclim2/lib/python3.10/site-packages/xcdat/regridder/accessor.py:91, in RegridderAccessor.grid(self)
     89 for axis in axis_names:
     90     try:
---> 91         data, bnds = self._get_axis_data(axis)
     92     except KeyError:
     93         continue

File ~/miniconda3/envs/xclim2/lib/python3.10/site-packages/xcdat/regridder/accessor.py:111, in RegridderAccessor._get_axis_data(self, name)
    106 def _get_axis_data(
    107     self, name: CFAxisKey
    108 ) -> Tuple[xr.DataArray | xr.Dataset, xr.DataArray]:
    109     coord_var = get_dim_coords(self._ds, name)
--> 111     _validate_grid_has_single_axis_dim(name, coord_var)
    113     try:
    114         bounds_var = self._ds.bounds.get_bounds(name, coord_var.name)

File ~/miniconda3/envs/xclim2/lib/python3.10/site-packages/xcdat/regridder/grid.py:744, in _validate_grid_has_single_axis_dim(axis, coord_var)
    725 """Validates that the grid's axis has a single dimension.
    726 
    727 If the grid has multiple dimensions (e.g., "lat" and "latitude" dims), xcdat
   (...)
    741     If the grid has multiple dimensions.
    742 """
    743 if isinstance(coord_var, xr.Dataset):
--> 744     raise ValueError(
    745         f"Multiple '{axis}' axis dims were found in this dataset, "
    746         f"{list(coord_var.dims)}. Please drop the unused dimension(s) before "
    747         "performing grid operations."
    748     )

ValueError: Multiple 'Z' axis dims were found in this dataset, ['plev7', 'tau']. Please drop the unused dimension(s) before performing grid operations.
```

### Environment

xCDAT version 0.6.1

INSTALLED VERSIONS
------------------
commit: None
python: 3.10.9 | packaged by conda-forge | (main, Feb  2 2023, 20:20:04) [GCC 11.3.0]
python-bits: 64
OS: Linux
OS-release: 3.10.0-1160.102.1.el7.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: ('en_US', 'UTF-8')
libhdf5: 1.12.2
libnetcdf: 4.8.1

xarray: 2023.4.2
pandas: 1.5.1
numpy: 1.23.4
scipy: 1.8.1
netCDF4: 1.6.1
pydap: None
h5netcdf: None
h5py: 3.8.0
Nio: None
zarr: None
cftime: 1.6.2
nc_time_axis: 1.4.1
PseudoNetCDF: None
iris: None
bottleneck: 1.3.5
dask: 2022.10.2
distributed: 2022.10.2
matplotlib: 3.6.2
cartopy: 0.21.0
seaborn: 0.13.1
numbagg: None
fsspec: 2022.11.0
cupy: None
pint: 0.20.1
sparse: 0.13.0
flox: None
numpy_groupies: None
setuptools: 65.5.1
pip: 22.3.1
conda: 23.9.0
pytest: None
mypy: None
IPython: 8.6.0
sphinx: None

**Comment by pochedls** [comment]
Unverified, but I think `cf-xarray` might be using the `tau` attribute `{"positive": "up"}` to determine that `tau` is a vertical axis. I think adding this resolves the issue:

```
del ds.tau.attrs['positive']
```

**Comment by mzelinka** [comment]
Thanks Steve, that worked.  Do you know why explicitly setting the coords / dims like I do in the example above does not work?

**Reply by pochedls** [reply]
I don't know. 

We all want xarray packages to recognize obvious axes (e.g., we should know that `lat` is latitude, whether it is labelled `axis = Y` or not). But inferring this kind of information creates grey areas (see this [recent example](https://github.com/xCDAT/xcdat/pull/602)). 

To *speculate*, it's possible that cf-xarray (or possibly xcdat) has created some functionality to tag axes with `"positive": "up"` as vertical coordinates. Someone should verify (probably by looking at this function `xc.get_dim_keys(ds, axis='Z')`)!

**Reply by mzelinka** [reply]
Thanks Steve

**Reply by tomvothecoder** [reply]
> To _speculate_, it's possible that cf-xarray (or possibly xcdat) has created some functionality to tag axes with `"positive": "up"` as vertical coordinates. Someone should verify (probably by looking at this function `xc.get_dim_keys(ds, axis='Z')`)!

cf_xarray interprets other CF attributes including `positive` and `long_name` ([link](https://cf-xarray.readthedocs.io/en/latest/coord_axes.html)).

xCDAT is supposed to use `cf-xarray` to interpret just the `axis` and `standard_name` attributes. In this case, xCDAT (with cf-xarray) is interpreting `positive`, which isn't our intended behavior.

We can open up a GitHub issue to narrow the scope of which attributes to interpret with cf-xarray.

**Reply by pochedls** [reply]
@tomvothecoder – can you clarify this – should xCDAT change how cf-xarray is used (I guess we can modify options when calling cf-xarray?) or are you saying we should open a github issue with cf-xarray to request / suggest changes?

**Reply by tomvothecoder** [reply]
> @tomvothecoder – can you clarify this – should xCDAT change how cf-xarray is used (I guess we can modify options when calling cf-xarray?) or are you saying we should open a github issue with cf-xarray to request / suggest changes?

The first option, although it might be tricky because we use `ds.cf.coords` (and `ds.cf.axes`) which I think interprets `positive` and the other attributes listed in that ([link](https://cf-xarray.readthedocs.io/en/latest/coord_axes.html)).

We can also live with it and maybe improve the error message somehow?

**Reply by pochedls** [reply]
It seems like we should improve the error message and link to the cf-xarray table you shared. I think this, because I suspect that changing this on xcdat's side will create more problems than it solves. Also, it seems like the dataset discussed in this thread shouldn't have `tau` labelled with `{positive: up}`. So maybe cf-xarray is doing the right thing?

**Reply by tomvothecoder** [reply]
> It seems like we should improve the error message and link to the cf-xarray table you shared. I think this, because I suspect that changing this on xcdat's side will create more problems than it solves. Also, it seems like the dataset discussed in this thread shouldn't have `tau` labelled with `{positive: up}`. So maybe cf-xarray is doing the right thing?

I agree, improving the error message and linking the cf-xarray table is a good idea. I'll open up an issue on it.

I think cf-xarray is doing the right thing here by interpreting `positive` for the Z axis. I'm not sure if `tau` is supposed to have that attribute set or not.

**Reply by tomvothecoder** [reply]
https://github.com/xCDAT/xcdat/issues/610
